Source/CarthageKitTests/Fakes/FakeURLSession.swift

```diff
@@ -0,0 +1,41 @@
+import Foundation
+
+final class FakeDownloadTask: NSURLSessionDownloadTask {
+	var started = false
+	override func resume() {
+		self.started = true
+	}
+
+	var cancelled = false
+	override func cancel() {
+		self.cancelled = true
+	}
+}
+
+final class FakeDataTask: NSURLSessionDataTask {
+	var started = false
+	override func resume() {
+		self.started = true
+	}
+
+	var cancelled = false
+	override func cancel() {
+		self.cancelled = true
+	}
+}
+
+final class FakeURLSession: NSURLSession {
```

Comment:
Is there a video for that talk? Would love to see the original talk, but the notes are really good information too.

> Now we can unit test createGitHubRequest, O.init for our object types, O.nextPage, and fetchAllPages doesn't need much testing. We can probably survive without testing it because it'll be obvious if it's completely broken, or we can add one or two integration-style tests (I would use an HTTP stubbing library for this).

So, essentially, you're proposing that we reduce the untested/untestable area of code to as small as possible. I'm a fan of that.

My main concern is having too many pure functions, then it becomes non-obvious how everything fits together. For example, when I want to show someone an example of terrible "clean" code, I show them `Xcode.swift`: Lots (boo!) of pure (yay!) functions, in addition to [this 200-line behemoth](https://github.com/Carthage/Carthage/blob/master/Source/CarthageKit/Xcode.swift#L868-L1043), and it's really hard to figure out what's going on because everything is (seemingly?) intermingled. And of course, with the way things are currently written, it is necessarily tested in an integration style.

> I'm going to add you as a contributor to Carthage. I think opening an issue for everything upfront will be overwhelming. I think a single giant issue would work (using a task list), or you could open up separate issues but not open the issue until you're nearly ready to work on it. That way you can get feedback, but you don't open 100 issues this week. :smile:

Awesome. Thank you! I opened https://github.com/Carthage/Carthage/issues/1147 to collect everything into an issue.